README.md
# Jarvis2.O
New and better console based personal assistant

</br>
</br>
</br>
</br>
</br>


# Website
https://www.letscodeofficial.com/jarvis


</br>
</br>
</br>
</br>
</br>

# Download
visit https://www.letscodeofficial.com/jarvis_downloads



</br>
</br>
</br>
</br>
</br>

# Updating 
You can update jarvis by either running this command inside jarvis

```
update jarvis
```

OR just downloading and installing the latest version


</br>
</br>
</br>
</br>
</br>


# Installing
</br>

## 1. Windows

visit - https://www.letscodeofficial.com/jarvis_downloadWindows2
</br>
</br>

## 2. Linux (Ubuntu/Debian)

visit - https://www.letscodeofficial.com/jarvis_downloadLinux2
</br>
</br>

You can run sudo jarvis command without entering password every time
</br>

just run this command in jarvis and follow instructions
```
add me to root
```
</br>
or run this command in terminal -

```shell
sudo gedit /etc/sudoers
```

then add this line to end of file
```
username ALL = (root) NOPASSWD: /bin/jarvis
```

remember to replace the username with your user name , you can find your username by running this command in terminal - 

```shell
whoami
```

</br>
</br>
</br>
</br>
</br>


# Online Documentation

visit - https://www.letscodeofficial.com/jarvis_docs_home


</br>
</br>
</br>
</br>
</br>


# Improvement to previous Gen Jarvis
link to previous gen jarvis - https://github.com/harshnative/JARVIS

from development point of view - 

1. Designed with modularity in mind that is all the internal function and packages returnes data to main function instead of priting is write away making them reusable anywere

2. better error handling and logging system

3. better caching and optimising in internal function

4. easily scalable

5. many more features



</br>
</br>
</br>
</br>
</br>

# Features

Jarvis 2.O can do many things such as

1. weather of any city
1. speed test
1. own settings file
1. password manager with military class encryption
1. Share file via both HTTP and FTP server with both wired and wireless medium.
1. mispelled command detection
1. auto complete command using tab
1. in built troubleshooter


More comming soon (^_^)



</br>
</br>
</br>
</br>
</br>


# OS suppport
Jarvis is currently desinged , tested and maintained for linux and windows only.

But project itself is made in python as has many other supported platform, making some changes the code should be able to make jarvis run on any operating system.


</br>
</br>
</br>
</br>
</br>

# Documentation
<a href="#Weather">1. Weather</a>
</br>
<a href="#Speed-Test">2. Speed Test</a>
</br>
<a href="#Settings">3. Settings</a>
</br>
<a href="#File-Share">4. File Share</a>
</br>
<a href="#Password-Manager">5. Password Manager</a>
</br>
<a href="#Auto-Completion">6. Auto Completion</a>
</br>
<a href="#troubleshoot">7. TroubleShoot</a>
</br>




</br>
</br>
</br>
</br>
</br>


<h2 id="Weather">
1. Weather
</h2>

```
weather cityName
```

Were cityName can be anything.
ex - london , new york


If no city name is passed then Jarvis check for the city name in city file

you can set the default city name in settings file.


</br>
</br>
</br>
</br>
</br>

<h2 id="Speed-Test">
2. Speed Test
</h2>

```
speed test
```

Will perform the speed test and average it 2 times by default to show you the most accurate real world result

</br>
</br>

```
speed test -b
```
If you want the result in mega bytes instead of mega bits

</br>
</br>

```
speed test 3
```
If you want the result to be averaged 3 times instead of 2

averaging is limited to max 5 times

</br>
</br>

```
speed test -b 3
```
You can combine the commands



</br>
</br>
</br>
</br>
</br>


<h2 id="Settings">
3. Settings
</h2>

```
open setting
```
To open the settings file

</br>
</br>

```
update setting
```
To apply the newly updates settings in Jarvis 


</br>
</br>

```
restore setting
```
To restore the settings file to default value 

</br>
</br>
Default settings file looks like - https://github.com/harshnative/Jarvis2.O/blob/main/defaultSettings.txt


</br>
</br>
</br>
</br>
</br>



<h2 id="File-Share">
4. File Share
</h2>

```
start file share
```
To start a ftp server
A window will be prompted to select the location to share

</br>
</br>
Options for start file share command - 

1. http
```
start file share http
```
To start a http server
usefull when you cannot access the ftp server as HTTP only require browser to work
Remember you can only download files using http server
</br>

2. -d
```
start file share -d
```
To start a server at default folder grabed from the settings file


3. -log
```
start file share -log
```
when you want to see realtime status of server
</br>

4. -a
```
start file share -a
```
when you want to allow anonymous access to FTP server , that is without ID password , but anonymous access is read only
</br>

5. command can be clubbed together
```
start file share http -d -log
```

now jarvis will start a http server with default dir from settings file in logging mode.

</br>
</br>


```
stop file share
```
To stop file share
</br>
</br>


```
set file share port PORTNUMBER
```
To set custom file share port if the default file share port is busy in your machine which is 5000

were PORTNUMBER can be from 1000 to 9999

when the port number assigned is busy , jarvis automatically chooses a free port
</br>
</br>


```
show file share
```
Show the output created by file share again to see the exact intructions to connect

</br>
</br>

#### username = user , password = 147896 - default user name and password , you can change these in settings file

#### Note - on linux you may not be able to access files shared via file share module due to permission issue you need to run this command to fix that 

```
add file to me
```

it will ask for username which you find by running whoami command in terminal

to avoid entering your user again and again you can use -d option to use username from settings file

This command basically running chown command on linux

</br>
</br>
</br>
</br>
</br>



<h2 id="Password-Manager">
5. Password Manager
</h2>

```
password
```
To start password manager

</br>

For the first time you will need to set the password for encryption.

password should be strong enough containing mixture of characters and min 8 chars in length.

You can also set a custom path to database in settings file (path should also contain the name of db like /user/mypasswords.db)

</br>

```
-a
```
To add a password
</br>
</br>


```
-a -m 
```
To add a multi line password

You will need to press on a new line - ctrl + d on linux or ctrl + z on windows to stop entering and add it
</br>
</br>

```
-u
```
To update a password
</br>
</br>




```
-d
```
To delete a password
</br>
</br>



```
-s
```
To see a password
</br>
</br>



```
-sa
```
To see all password 
</br>
</br>


```
-c
```
To change password 
</br>
</br>



```
-sa all
```
Jarvis has started to crush the long references and passwords while showing them in -sa to maintian a cleaner look

To see all the characters in reference and passwords you can add all to the command

like 
-sa all
-u all 
-d all

</br>
</br>
</br>
</br>
</br>

<h2 id="Auto-Completion">
6. Auto-Completion
</h2>

After writing some part of the command you can press tab to auto complete it

Auto completion does not work when entering variables.

You commands are automatically corrected when mispelled 




</br>
</br>
</br>
</br>
</br>


<h2 id="troubleshoot">
7. Troubleshoot
</h2>

```
troubleshoot
```

run this command to start the trouble shooter and follow the instructions

Troubleshooter can the end will upload the log file to server for developers to investigate , it does not contain any sensitive information and password module does not have looger.

If you participated in User experience programming then you dont need to re run the command , log file will directly be uploaded

you can always opt in or out of UEP in settings file

</br>
</br>
</br>
</br>
</br>


# Some backend details

1. Storage directory : On windows jarvis stores user file such as logs , settings , data bases etc in C:\programData\JarvisData. While in ubuntu it stores in /opt/JarvisData.

2. Encryption used in password Data base - Jarvis uses military class encryption technique provided by easySED module to securely store your passwords. you learn more about easySED at https://www.letscodeofficial.com/pySEDDocs



</br>
</br>
</br>
</br>
</br>


# Contributors
<a href="https://www.letscodeofficial.com/"><img src="https://www.letscodeofficial.com/static/images/favicon.ico" width="150" height="150" /><h3>Lets Code Official</h3></a>


</br>
</br>
</br>

<a href="https://github.com/harshnative/"><img src="https://www.letscodeofficial.com/static/images/jarvis/HarshNativeProfile.JPG" width="150" height="150" /><h3>Harsh Native</h3></a>


More contributors are listed on website - https://www.letscodeofficial.com/jarvis_contributors



</br>
</br>
</br>
</br>
</br>


# Donate 🥰
You can donate at https://www.letscodeofficial.com/jarvis_contribute


</br>
</br>
</br>
</br>
</br>


# Contribute
Comming soon 



</br>
</br>
</br>
</br>
</br>


# license
Jarvis is licensed under GNU GENERAL PUBLIC LICENSE VERSION 3

Visit www.letscodeofficial.com/gnuV3 for license terms



changeLog.md
# v0.3
1. -nolog option for fileshare

2. auto file add option in settings for file share

3. start file share -n verbose mode also option in settings for file share

4. stores previously executed commands even after machine shutdown

5. show suggestions for the command to run

6. improved auto completion

</br>
</br>
</br>
</br>
</br>

# v0.2
1. fixed jarvis data folder system in linux

2. fixed weather city not found bug

3. added version checker API

4. added password copy to clipboard functionality in password module

5. added more pretty display of tabular data in password module using pandas

6. added table formatting to tabulate in password module

7. added update jarvis command

8. auto complete command feature

9. jarvis remembers previously runned command outside inner modules

10. now jarvis can also understand mispelled commands

11. password module now as skip option in update command as well

12. password module now as delete all option as well

13. improved interaction with password module

14. user can participate in the user experience program

15. user can now add multi line input in password module

16. user can set a default directory for file share

17. file share now supports displaying log messages to console



</br>
</br>
</br>
</br>
</br>


# v0.1

1. Designed with modularity in mind that is all the internal function and packages returnes data to main function instead of priting is write away making them reusable anywere

2. better error handling and logging system

3. better caching and optimising in internal function

4. easily scalable

main.py
# essential modules
import os
import platform
import time
import sys
import subprocess as sp
from threading import Thread
sys.path.append(os.path.join(os.path.dirname(__file__), '.'))
from packages.logger import loggerpy 
import multiprocessing




# declaring some global variables
class GlobalData_main:

    # variables to determine the operating system of the user 
    isOnWindows = False
    isOnLinux = False

    # variables to manage the loading animation 
    runLoadingAnimation = True
    loadingAnimationCount = 5
    lAnimationObj = None


    # path for storing the program files
    folderPathWindows = r"C:\programData\JarvisData"
    folderPathLinux = os.getcwd() + "/JarvisData"
    folderPathWindows_simpleSlash = r"C:/programData/JarvisData"

    # obj for logger module
    objClogger = None
    troubleshootValue = True

    # dict from the settings file will be stored here
    settingDict = None


    # variable for file share module
    portForFileShare = 5000 
    isFileShareStarted = False
    addressForShare = None
    dataListFileShare = []

    # obj to store tk object
    root = None

    # list holding info about running background operations
    printDataList = []































# Checking the users operating system and adding data to global class
osUsing = platform.system()

if(osUsing == "Linux"):
    GlobalData_main.isOnLinux = True

    # making the program data folder
    try:
        os.mkdir(GlobalData_main.folderPathLinux)
    except FileExistsError:
        pass

elif(osUsing == "Windows"):
    GlobalData_main.isOnWindows = True

    # making the program data folder
    try:
        os.mkdir(GlobalData_main.folderPathWindows)
    except FileExistsError:
        pass

else:
    print("Jarvis currently does not support this operating system :(")
    time.sleep(3)
    sys.exit()

del osUsing































# clear screen function 
def customClearScreen():
    if(GlobalData_main.isOnWindows == True):
        os.system("cls")
    else:
        sp.call('clear',shell=True)








































# setting up the logger module
GlobalData_main.objClogger = loggerpy.Clogger(GlobalData_main.isOnWindows , GlobalData_main.isOnLinux , GlobalData_main.folderPathWindows_simpleSlash , GlobalData_main.folderPathLinux)
GlobalData_main.objClogger.setTroubleShoot(GlobalData_main.troubleshootValue)





# loading animation thread
class LoadingAnimation(Thread):

    def run(self):

        customClearScreen()
        string = ""

        # will run only till max 5 times or early if the global var runLoadingAnimation is made false
        while(GlobalData_main.runLoadingAnimation and GlobalData_main.loadingAnimationCount):
            string = string + "."
            time.sleep(0.5)
            print("\rloading , please wait " , string , sep = "" , end = "")
            GlobalData_main.loadingAnimationCount -= 1

        print()






# starting in the main to avoide awakeness by sub processes
if __name__ == "__main__":
    multiprocessing.freeze_support()

    # loading animation thread started 
    GlobalData_main.lAnimationObj = LoadingAnimation()
    GlobalData_main.lAnimationObj.start()




























# importing additional modules
from easyOpenWeather import module as owm
from tabulate import tabulate

from packages.speedTest import speedTestFile
from packages.settingM import settingsFile
from packages.passwordStorer import mainCode



# importting components for file share functionality
from packages.fileShare import FS
from tkinter import filedialog
from tkinter import *

if __name__ == "__main__":
    GlobalData_main.root = Tk()
    GlobalData_main.root.withdraw()






























# class containing some of the global methods
class GlobalMethods:

    # method for comparing if a string as a certain substring
    @classmethod
    def isSubString(cls , string , subString):
        lengthOfSubString = len(subString)
        try:
            """ for substring to be string - 
                1. check if the first of substring matches the string as certain i position
                2. for that i , check wheather the next characters also matches the substring till i + length of substring"""

            for i, j in enumerate(string):
                if(j == subString[0]):
                    if(subString == string[i:i+lengthOfSubString]):
                        
                        # logging the result for debugging purpose
                        GlobalData_main.objClogger.log("subString found in GlobalMethod class isSubString function , values(string , subString) = ({} , {})".format(string , subString) , "i")
                        return True
                    else:
                        pass

            # logging the result for debugging purpose
            GlobalData_main.objClogger.log("subString not found in GlobalMethod class isSubString function , values(string , subString) = ({} , {})".format(string , subString) , "i")
            return False

        except Exception as e:

            # logging the error
            GlobalData_main.objClogger.exception(str(e) , "Error in GlobalMethod class isSubString function , values(string , subString) = ({} , {})".format(string , subString))
            return False



    # method for checking if the substring is in string using list
    @classmethod
    def isSubStringsList(cls , string , subString):

        """ This method check is this way :
            normal substring check - string(hello world , this is jarvis) , subString(this jarvis) -> False
            using list substring check - string(hello world , this is jarvis) , subString(this jarvis) -> True
            compares the invidual words as substring"""

        # converting them to same case
        string = string.upper()
        subString = subString.upper()

        # getting the list of words in substring sperated by space
        subStringList = subString.split()

        # vars to keep track of number of successfull comparisions
        count1 = 0
        count2 = 0

        for i in subStringList:
            # striping the leading and trailing spaces of word in sub string list
            i = i.strip()

            # increasing the count1 as we have traversed a word
            count1 += 1

            # if this word is also a substring then we increase the count2 as well
            if(cls.isSubString(string , i)):
                count2 += 1
            else:
                # returned false as now substring is not in string
                return False

        # if the count1 matches the count2 and count1 should be greater than 0 then substring is in string
        if((count1 == count2) and count1 > 0):
            return True
        else:
            return False






































# class to get the wheather details
class WeatherFunctionality:
    
    def __init__(self):
        
        # making the object
        self.moduleObj = owm.WeatherDataClass()

        # set the api key here
        # it is removed before commiting for security reasons
        self.__apiKey = "fe82651e607e46db61dba45e39aa7e17"

        # if the custom api key is not setted then the default api key present with the module will be used
        if(self.__apiKey != ""):
            self.moduleObj.setApiKey(self.__apiKey)
    

    # driver function
    def returnDataDict(self , cityName):

        # setting the city name as per https://github.com/harshnative/easyOpenWeather_module
        self.moduleObj.setCityName(cityName)
    
        # these the things for which the data will be retreived
        listPass = ["tempInC" , "tempMin" , "tempMax" , "pressure" , "humidity" , "windSpeed" , "windDirection" , "clouds" , "description"]
        self.moduleObj.setList(listPass)


        try:
            # returning the info in dict form
            finalResult = self.moduleObj.getInfo()
            
            # logging the result for debugging purpose
            GlobalData_main.objClogger.log("weather details retrieved successfully in weather functionality class returnDataDict function , values(cityName) = ({})".format(cityName) , "i")

            return finalResult

        except Exception as e:

            # logging the error
            GlobalData_main.objClogger.exception(str(e) , "Exception in getting the weather details in WeatherFunctionality class and returnDataDict function , values(cityName) = ({})".format(cityName))
            
            return "error , could not get the weather details :("

































# class for settings functionality
class Settings:

    # setting up the object 
    settingObj = settingsFile.SettingsClass(GlobalData_main.folderPathWindows_simpleSlash , GlobalData_main.folderPathLinux , GlobalData_main.isOnWindows , GlobalData_main.isOnLinux)

    # method to return the dict
    @classmethod
    def returnDict(cls):
        returnedDict = cls.settingObj.getDict()
        return returnedDict

    # method to open the settings file
    # returns True on successfull opening 
    # else returns false and logs error
    @classmethod
    def openSettingsFile(cls):
        result = cls.settingObj.openSettings()

        if(result == None):
            return True
        else:
            GlobalData_main.objClogger.exception(str(result) , "Exception in opening the settings file")
            return False

    
    # method to restore the settings file with default settings
    @classmethod
    def restoreSettings(cls):
        cls.settingObj.regenerateSettingsFile()






















class Help:
    pass
























# class to implement the file share functionality
class FileShareClass:


    # some class vars
    fil = FS.FileShareClass()
    path = None 

    
    # function to set the share path for the file share
    # open the tk gui window and then stores the returned path in class var
    # return False if process fails else return True
    @classmethod
    def setSharePath(cls):
        try:
            folder_selected = filedialog.askdirectory()
        except Exception as e:
            GlobalData_main.objClogger.exception(str(e) , "Exception in getting folder path from tkinter window")
            return False

        cls.path = folder_selected
        return True




    # function to start the file share 
    # returns None in case of process failure
    # else returns a list of output from module
    @classmethod
    def startFileShare(cls , http = False):
        try:
            result = cls.fil.start_fileShare(cls.path , http=http , port=GlobalData_main.portForFileShare)
            GlobalData_main.dataListFileShare = result
            GlobalData_main.addressForShare = cls.fil.get_ip_address()
            GlobalData_main.printDataList.append("File Share active at {}:{}".format(str(GlobalData_main.addressForShare) , str(GlobalData_main.portForFileShare)))
            GlobalData_main.objClogger.log("file share server started at {}".format(str(cls.path)) , "i")
            GlobalData_main.isFileShareStarted = True
            return result
        except Exception as e:
            GlobalData_main.objClogger.exception(str(e) , "Exception in starting the file share server at {}".format(str(cls.path)))
            return None

    
    # function to stop the file share module
    # return True if process completes
    # else returns False
    @classmethod
    def stopFileShare(cls):
        try:
            cls.fil.stopFileShare()
            cls.fil = FS.FileShareClass()
            GlobalData_main.printDataList.clear()
            try:
            	GlobalData_main.printDataList.remove("File Share active at {}:{}".format(str(GlobalData_main.addressForShare) , str(GlobalData_main.portForFileShare)))
            except Exception as e:
            	GlobalData_main.printDataList.clear()
            GlobalData_main.isFileShareStarted = False
            GlobalData_main.objClogger.log("file share server stopped at {}".format(str(cls.path)) , "i")
            return True
        except Exception as e:
            GlobalData_main.objClogger.exception(str(e) , "Exception in stopping the file share server at {}".format(str(cls.path)))
            return False





























# main driver function for executing commands
def driver(command):

    """No print is used before main to keep the functions modular , just return the result to main in list form or yield form"""
    

    # if the weather command is passed
    if(GlobalMethods.isSubStringsList(command , "weather")):

        # getting the default city name
        cityNameFromSet = GlobalData_main.settingDict.get("cityName" , "None")

        # getting the city name from command(weather cityName)
        cityName = command[len("weather") + 1:]

        cityName = cityName.strip()
        
        # if the city name is None is settings file and also city name is not passed in command then error will be returned to user
        if((cityNameFromSet == "None") and (len(cityName) <= 1)):
            yield "please pass the city name like weather london or set the default city name in settings file"
            yield "\nYou can open settings file by typing settings command"
            return

        # else if city is not passed then it will be taken from the settings file
        elif((len(cityName) <= 1)):
            cityName = cityNameFromSet
            

        # getting the result from the class
        weatherObj = WeatherFunctionality()

        result = weatherObj.returnDataDict(cityName)

        # if the api key is not set then return the message
        if(result == None):
            yield "please set the api key"
            return
        
        if(result == "error"):
            yield "please pass a correct city name"
            return

        # else conv the dict returned into list with [[key] , [value]]
        resultList = []

        for i , j in result.items():
            tempList = []
            tempList.append(i)
            tempList.append(j)
            resultList.append(tempList)

        # using the tabulate module to pretify the output
        toReturn = tabulate(resultList , headers=['Query', 'Data'])

        # returning the result so it can be printed
        yield toReturn



















    # if the speed test  command is passed
    if(GlobalMethods.isSubStringsList(command , "speed test")):
        sObj = speedTestFile.SpeedTestClass(GlobalData_main.objClogger)

        # spliting the command to get command list
        commandList = command.split()

        inBytes = False
        numberOfTimes = 2

        # if user wants result in bytes instead of bits
        if(("-b" in commandList) or ("-B" in commandList)):
            inBytes = True

        # finding the user as passed number of times
        for i in commandList:
            if(i.isnumeric()):
                if(int(i) <= 5):
                    numberOfTimes = int(i)
        


        for i in sObj.runSpeedTestUtility(inBytes=inBytes , numberOfTimesToDo=numberOfTimes):
            yield i
        
        return





















    # if the setting command is passed
    if(GlobalMethods.isSubStringsList(command , "setting")):

        # if update setting command is passed
        if(GlobalMethods.isSubStringsList(command , "update")):
            GlobalData_main.settingDict = Settings.returnDict()
            yield "new settings applied ..."
            return
        
        # if update setting command is passed
        if(GlobalMethods.isSubStringsList(command , "restore")):
            Settings.restoreSettings()
            yield "settings restored to default values ..."
            return

        # if update setting command is passed
        if(GlobalMethods.isSubStringsList(command , "open")):
            if(Settings.openSettingsFile()):
                yield "Settings file opened"
                yield "\nDon't forget to save it and run the update command after wards"
                return
            else:
                yield "Error opening the settings file , Try running the troubleshoot command"
                return























    # if the password command is passed
    if(GlobalMethods.isSubStringsList(command , "password")):

        # getting the path from settings dict
        customPath = GlobalData_main.settingDict.get("pathForPassDB")


        # if path is not setted then simply create new db
        if(customPath.lower() == "none"):
            customPath = None
        else:

            # if the path is setted but is incorrect or jarvis does not have permission to read write at that file dir then simply promt a message
            if(not(os.path.isfile(customPath))):
                print("Path specified in settings file for \"pathForPassDB\" is either incorrect or jarvis does not permission to read write their.\npassword module will create a new data base now")

                inputted = input("\n\nEnter 1 to continue or anything else to quit")
                
                if(inputted != 1):
                    yield "clear screen"
                    yield "cannot open password db"
                    return True
                    

        # waking the driver function of password module
        obj = mainCode.PasswordStorerClass(GlobalData_main.isOnWindows , GlobalData_main.isOnLinux , GlobalData_main.folderPathWindows_simpleSlash , GlobalData_main.folderPathLinux , customPath)
        obj.driverFunc()

        yield "clear screen"
        yield "exitted password db"
        return True
    




















    # if set port command is passed
    if(GlobalMethods.isSubStringsList(command , "set file port")):

        # splitting the command string to get a list
        commandList = command.split()

        port = 5000

        # traversing the command list until a numeric value is found as i
        # then set that numeric value as port
        for i in commandList:
            i = str(i)
            if(i.isnumeric()):
                port = i
            
        port = int(port)

        # port must be btw 1000 and 9999
        if((port < 1000) or (port > 9999)):
            yield "port number must be a four digit integer"
            return True
        
        # setting the port number to global var
        GlobalData_main.portForFileShare = int(port)
        yield "port number setted to {}, now you can start the file share with this port".format(str(port))
        return True






    # if stop file share command is passed
    if(GlobalMethods.isSubStringsList(command , "stop file")):

        if(GlobalData_main.isFileShareStarted):
            status = FileShareClass.stopFileShare()

            if(not(status)):
                yield "fileShare module could not stop the process"
                return True
        
            yield "filShare Stopped successfully"
            return True
        
        else:
            yield "file share is not running"
            return True



    # if the start file share command is passed
    if(GlobalMethods.isSubStringsList(command , "start file")):

        status = FileShareClass.setSharePath()

        if(not(status)):
            yield "fileShare module could not get path, Try again"
            return True

        http = False

        if(GlobalMethods.isSubStringsList(command , "http")):
            http = True

        result = FileShareClass.startFileShare(http=http)

        if(result == None):
            yield "fileShare module could not start the server, Try again"
            return True

        yield "clear screen"
        for i in result:
            yield str(i)
            yield "\n"

        return True

    # if show file share command is passed
    if(GlobalMethods.isSubStringsList(command , "show file")):

        yield "clear screen"
        
        for i in GlobalData_main.dataListFileShare:
            yield str(i)
            yield "\n"

        return True
        





























    # if exit all command is passed
    if(GlobalMethods.isSubStringsList(command , "exit all")):
        if(GlobalData_main.isFileShareStarted):
            try:
                FileShareClass.stopFileShare()
            except Exception as e:
                GlobalData_main.objClogger.exception(str(e) , "Exception in closing the file share server at {} while exit all".format(str(FileShareClass.path)))

        sys.exit()


    # if exit command is passed
    if(GlobalMethods.isSubStringsList(command , "exit")):
        if(GlobalData_main.isFileShareStarted):
            yield "could not quit the program, file share is already running, use stop file share to quit it first"
            return True

        sys.exit()


    


    else:
        yield "Command not found , Try again , or Type Help for Help"

























# main function for inputting commands
def main():

    # running until exit is called in driver
    while(True):
        customClearScreen()

        userName = GlobalData_main.settingDict.get("username" , "sir")

        for i in GlobalData_main.printDataList:
            print(i)
            print()

        # greeting the user
        print("Welcome {} , What can i do for you :)\n".format(userName))

        # inputting the command
        command = input("Enter Command : ")

        customClearScreen()

        # priting the result from driver function
        for i in driver(command):
            if(i == "clear screen"):
                customClearScreen()
            else:
                print(i)

        print("\n\nPress Enter to continue ...")
        input()


        




# main main 
if __name__ == "__main__":
    
    multiprocessing.freeze_support()

    # closing the loading animation
    GlobalData_main.runLoadingAnimation = False
    GlobalData_main.lAnimationObj.join()

    # A message will be printed if the troubleshoot value is True by default
    if(GlobalData_main.troubleshootValue):
        print("\nIn dev Mode \n")
        time.sleep(0.5)

    customClearScreen()

    # cache the settings dict in main memory
    try:
        GlobalData_main.settingDict = Settings.returnDict()
    
    # if the cache fails then the settings file retored may not have been written due to not availabilty of cpu
    except Exception as e:
        GlobalData_main.objClogger.exception(str(e) , "Exception in getting the dict from the settings file , may be the cpu was busy , trying again in 1 sec")
        time.sleep(1)

        # after one second jarvis will try again
        try:
            GlobalData_main.settingDict = Settings.returnDict()

            # if jarvis fails this time also then the computer must have been in deadlock state or jarvis does not have read write permission to program's data folder
        except Exception as e:
            GlobalData_main.objClogger.exception(str(e) , "Exception in getting the dict from the settings file , may be the cpu was busy , or jarvis does not have read write permission")
            customClearScreen()
            print("Could not load the settings file , cpu does not responded or jarvis does not have read write permission , Try again in some time :(")
            print("\n\nPress enter to continue")
            input()
            sys.exit()


    # main will be called
    main()
